extensions/amp-date-display/0.2/date-display.js

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {useResourcesNotify} from '../../../src/preact/utils';
+import {userAssert} from '../../../src/log';
+
+/** @const {string} */
+const DEFAULT_LOCALE = 'en';
+
+/** @const {number} */
+const DEFAULT_OFFSET_SECONDS = 0;
+
+/** @typedef {{
+  year: number,
+  month: number,
+  monthName: string,
+  monthNameShort: string,
+  day: number,
+  dayName: string,
+  dayNameShort: string,
+  hour: number,
+  minute: number,
+  second: number,
+  iso: string,
+}} */
+let VariablesV2Def;
+
+/** @typedef {{
+  year: number,
+  month: number,
+  monthName: string,
+  monthNameShort: string,
+  day: number,
+  dayName: string,
+  dayNameShort: string,
+  hour: number,
+  minute: number,
+  second: number,
+  iso: string,
+  yearTwoDigit: string,
+  monthTwoDigit: string,
+  dayTwoDigit: string,
+  hourTwoDigit: string,
+  hour12: string,
+  hour12TwoDigit: string,
+  minuteTwoDigit: string,
+  secondTwoDigit: string,
+  dayPeriod: string,
+ }} */
+let EnhancedVariablesV2Def;
+
+/**
+ * @param {!JsonObject} props
+ * @return {Preact.Renderable}
+ */
+export function DateDisplay(props) {
+  const render = props['render'];
+  const data = /** @type {!JsonObject} */ (getDataForTemplate(props));
+  useResourcesNotify();
```

Comment:
I'm still struggling this this call. I feel like if we constraint it strictly to only be in the `RenderDomTree` that should be sufficient? Then we have a nice separation of concerns: `DateDisplay` functions more or less strictly within the React-style. Whole `RenderDomTree` adopts it to AMP.

Reply:
I think this is ok to start this way. But I'll still be asking and looking for something more composite here. We have pretty clear touchpoints where we deal with actual light/shadow DOMs between `Slot` and `RenderDomTree`. Also, per our React prototype, it's very likely that we will decide to wrap all/most of AMP elements into some form of a `Root` component to preserve the common layout semantics. The bottom line: let's roll with it for now, but we'll likely come back to it again.